### PR TITLE
Track EOS for each connected receiver to the NetworkPublication

### PR DESCRIPTION
**Pull request by mikeb01**
Disconnect if all have responded with an EOS.

### CODE REVIEW

File: aeron-driver/src/main/java/io/aeron/driver/NetworkPublication.java [hunk 1]
```diff
@@ -714,9 +726,12 @@ boolean hasSpies()
 
     void updateHasReceivers(final long timeNs)
     {
-        if (((timeOfLastStatusMessageNs + connectionTimeoutNs) - timeNs < 0) && hasReceivers)
+        livenessTracker.onIdle(timeNs, connectionTimeoutNs);
+        final boolean isLive = livenessTracker.hasReceivers();
+
+        if (hasReceivers != isLive)
```
Comment (vyazelenko) [inline aeron-driver/src/main/java/io/aeron/driver/NetworkPublication.java:732]: Why conditional assignment, i.e. volatile write?

File: aeron-system-tests/src/test/java/io/aeron/SubscriberEndOfStreamTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.lessThan;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class SubscriberEndOfStreamTest
```
Comment (vyazelenko) [inline aeron-system-tests/src/test/java/io/aeron/SubscriberEndOfStreamTest.java:40]: Should be annotated with `@ExtendWith({ EventLogExtension.class, InterruptingTestCallback.class })`.

File: aeron-system-tests/src/test/java/io/aeron/SubscriberEndOfStreamTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    @AfterEach
+    void tearDown()
+    {
+        CloseHelper.quietClose(driver);
+    }
+
+    private TestMediaDriver launch()
+    {
+        driver = TestMediaDriver.launch(context, systemTestWatcher);
+        return driver;
```
Comment (vyazelenko) [inline aeron-system-tests/src/test/java/io/aeron/SubscriberEndOfStreamTest.java:61]: Should register `aeron.dir` with data collector, i.e.
```java
systemTestWatcher.dataCollector().add(driver.context().aeronDirectory());
```

**Review by vyazelenko — CHANGES_REQUESTED** [review]

File: aeron-driver/src/main/java/io/aeron/driver/NetworkPublication.java (hunk 1 above)
Comment (mikeb01) [reply aeron-driver/src/main/java/io/aeron/driver/NetworkPublication.java:732]: Yes, it is to prevent the cost of the volatile write when not required.

**Review by vyazelenko — APPROVED** [review]

**Comment by vyazelenko** [comment]
LGTM

File: aeron-driver/src/main/java/io/aeron/driver/ReceiverLivenessTracker.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package io.aeron.driver;
+
+import org.agrona.collections.Long2LongHashMap;
+
+class ReceiverLivenessTracker
+{
+    private static final int MISSING_VALUE = -1;
+
+    private final Long2LongHashMap liveMap = new Long2LongHashMap(MISSING_VALUE);
```
Comment (mjpt777) [inline aeron-driver/src/main/java/io/aeron/driver/ReceiverLivenessTracker.java:24]: liveMap -> livenessTimestampByReceiverIdMap?

**Pull request merged**